readme.md
# Industrypicker

> A simple jQuery plugin for picking the Industry kinds of China.

- [Homepage](http://lizhaoch.github.io/IndustyPicker)



## Table of contents

  - [Main](#main)
  - [Getting started](#getting-started)
  - [Options](#options)
  - [Methods](#methods)
  - [No conflict](#no-conflict)
  - [Browser support](#browser-support)
  - [License](#license)


## Main

```
dist/
├── Industrypicker.js          ( 7 KB)
├── Industrypicker.data.js     (97 KB)
```

## Getting started


### Quick start

Four quick start options are available:

- [Download the latest release](https://github.com/lizhaoch/IndustyPicker/archive/master.zip).
- Clone the repository: `git clone https://github.com/lizhaoch/Industrypicker.git`.


### Installation

Include files:

```html
<script src="/path/to/jquery.js"></script><!-- jQuery is required -->
<script src="/path/to/Industrypicker.data.js"></script>
<script src="/path/to/Industrypicker.js"></script>
```


Create HTML elements:

```html
<div><!-- container -->
  <select></select><!-- 门类 -->
  <select></select><!-- 大类 -->
  <select></select><!-- 中类 -->
  <select></select><!-- 小类 -->
</div>
```



### Usage

#### Initialize with `data-toggle="Industrypicker"` attribute


Basic

```html
<div data-toggle="Industrypicker">
  <select></select>
  <select></select>
  <select></select>
  <select></select>
</div>
```

#### Initialize with `$.fn.Industrypicker` method

Basic

```js
$('#target').Industrypicker();
```

Custom placeholders

```js
$('#target').Industrypicker({
  menlei: '—— 门类 ——',
        dalei: '—— 大类 ——',
        zhonglei: '—— 中类 ——',
        xiaolei: '—— 小类 ——'
});
```

[⬆ back to top](#table-of-contents)



## Options

- Change the default options with `$().Industrypicker(options)`.
- Change the global default options with `$.fn.Industrypicker.setDefaults(options)`.


### autoSelect

- Type: `Boolean`
- Default: `true`

Selects the city and district automatically when the province changes.


### placeholder

- Type: `Boolean`
- Default: `true`

Show placeholder (with an `<option>` element).


### menlei门类

- Type: `String`
- Default: `—— 门类 ——`

Defines the initial value of province `<select>`. If it is a existing province in `Industrypicker.data.js`, it will be selected. If not, it will be used as a placeholder.


### dalei大类

- Type: `String`
- Default: `—— 大类 ——`

Defines the initial value of city `<select>`. If it is a existing city under the selected province, it will be selected. If not, it will be used as a placeholder.


### zhonglei中类

- Type: `String`
- Default: `—— 中类 ——`

### xiaolei小类

- Type: `String`
- Default: `—— 小类 ——`



Defines the initial value of district `<select>`. If it is a existing district under the selected city, it will be selected. If not, it will be used as a placeholder.

[⬆ back to top](#table-of-contents)

## Methods

### reset([deep])

- **deep** (optional):
  - Type: `Boolean`
  - Default: `false`
  - Reset the selects to default states (Undo selected).

Reset the selects to the initial states (Undo changed).

**Examples:**

```js
$().Industrypicker('reset');
$().Industrypicker('reset', true);
```

### destroy()

Destroy the Industrypicker instance, but keep the selected districts.

If you want to remove the selected districts, you can call `reset` method first and then call this method.


[⬆ back to top](#table-of-contents)

## Browser support 浏览器支持

- Chrome (latest)
- Firefox (latest)
- Safari (latest)
- Opera (latest)
- Edge (latest)
- Internet Explorer 8+

As a jQuery plugin, you also need to see the [jQuery Browser Support](http://jquery.com/browser-support/).



## License

[MIT](http://opensource.org/licenses/MIT) © LIZHAOCH


[⬆ back to top](#table-of-contents)


CHANGELOG.md
# Changelog


### 0.0.1 (Jun 17, 2016)

- Updated Industry data.

js/industrypicker.data.js
/*!
 * industrypicker v1.0.4
 * https://github.com/fengyuanchen/industrypicker
 *
 * Copyright (c) 2014-2016 Fengyuan Chen
 * Released under the MIT license
 *
 * Date: 2016-06-01T15:05:52.606Z
 */

(function (factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as anonymous module.
    define('ChineseIndustry', [], factory);
  } else {
    // Browser globals.
    factory();
  }
})(function () {

  var ChineseIndustry = {
    86: {
      A: '农、林、牧、渔业',
      B: '采矿业',
      C: '制造业',
      D: '电力、热力、燃气及水生产和供应业'
    },
    A: {
      01: '农业',
      02: '林业',
      03: '畜牧业'
    },
    01: {
      011: '谷物种植',
      012: '豆类、油料和薯类种植'
    },
    011: {
      0111: '稻谷种植',
      0112: '小麦种植'
    },
    012: {
      0121: '豆类种植',
      0122: '油料种植'
    },
    B: {
      06: '煤炭开采和洗选业',
      07: '唐山市'
    },
    06: {
      061: '烟煤和无烟煤开采洗选',
      062: '褐煤开采洗选'
    },
    061: {
      0610: '褐煤开采洗选'
    },
    062: {
      0620: '烟煤和无烟煤开采洗选'
    },
    07: {
      071: '石油开采',
      072: '天然气开采',
    },
    071: {
      0710: '石油开采'
    },
    072: {
      0720: '天然气开采'
    }
  };

  if (typeof window !== 'undefined') {
    window.ChineseIndustry = ChineseIndustry;
  }
  return ChineseIndustry;
});


javascripts/industrypicker.js
/*!
 * industrypicker v1.0.4
 * https://github.com/fengyuanchen/industrypicker
 *
 * Copyright (c) 2014-2016 Fengyuan Chen
 * Released under the MIT license
 *
 * Date: 2016-06-01T15:05:52.606Z
 */

(function (factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as anonymous module.
        define(['jquery', 'ChineseIndustry'], factory);
    } else if (typeof exports === 'object') {
        // Node / CommonJS
        factory(require('jquery'), require('ChineseIndustry'));
    } else {
        // Browser globals.
        factory(jQuery, ChineseIndustry);
    }
})(function ($, ChineseIndustry) {

    'use strict';

    if (typeof ChineseIndustry === 'undefined') {
        throw new Error('The file "industrypicker.data.js" must be included first!');
    }

    var NAMESPACE = 'industrypicker';
    var EVENT_CHANGE = 'change.' + NAMESPACE;
    var menlei = 'menlei';
    var dalei = 'dalei';
    var zhonglei = 'zhonglei';
    var xiaolei = 'xiaolei';

    function industrypicker(element, options) {
        this.$element = $(element);
        this.options = $.extend({}, industrypicker.DEFAULTS, $.isPlainObject(options) && options);
        this.placeholders = $.extend({}, industrypicker.DEFAULTS);
        this.active = false;
        this.init();
    }

    industrypicker.prototype = {
        constructor: industrypicker,

        init: function () {
            var options = this.options;
            var $select = this.$element.find('select');
            var length = $select.length;
            var data = {};

            $select.each(function () {
                $.extend(data, $(this).data());
            });

            $.each([menlei, dalei, zhonglei, xiaolei], $.proxy(function (i, type) {
                if (data[type]) {
                    options[type] = data[type];
                    this['$' + type] = $select.filter('[data-' + type + ']');
                } else {
                    this['$' + type] = length > i ? $select.eq(i) : null;
                }
            }, this));

            this.bind();

            // Reset all the selects (after event binding)
            this.reset();

            this.active = true;
        },

        bind: function () {
            if (this.$menlei) {
                this.$menlei.on(EVENT_CHANGE, (this._changemenlei = $.proxy(function () {
                    this.output(dalei);
                    this.output(zhonglei);
                    this.output(xiaolei);
                }, this)));
            }

            if (this.$dalei) {
                this.$dalei.on(EVENT_CHANGE, (this._changedalei = $.proxy(function () {
                    this.output(zhonglei);
                    this.output(xiaolei);
                }, this)));
            }

            if (this.$zhonglei) {
                this.$zhonglei.on(EVENT_CHANGE, (this._changezhonglei = $.proxy(function () {
                    this.output(xiaolei);
                }, this)));
            }
        },

        unbind: function () {
            if (this.$menlei) {
                this.$menlei.off(EVENT_CHANGE, this._changemenlei);
            }

            if (this.$dalei) {
                this.$dalei.off(EVENT_CHANGE, this._changedalei);
            }
        },

        output: function (type) {
            var options = this.options;
            var placeholders = this.placeholders;
            var $select = this['$' + type];
            var zhongleis = {};
            var data = [];
            var code;
            var matched;
            var value;

            if (!$select || !$select.length) {
                return;
            }

            value = options[type];
            code = (
                type === menlei ? 86 :
                    type === dalei ? this.$menlei && this.$menlei.find(':selected').data('code') :
                        type === zhonglei ? this.$dalei && this.$dalei.find(':selected').data('code') :
                            type === xiaolei ? this.$zhonglei && this.$zhonglei.find(':selected').data('code') : code
            );

            zhongleis = ChineseIndustry[code];

            if ($.isPlainObject(zhongleis)) {
                $.each(zhongleis, function (code, address) {
                    var selected = address === value;

                    if (selected) {
                        matched = true;
                    }

                    data.push({
                        code: code,
                        address: address,
                        selected: selected
                    });
                });
            }

            if (!matched) {
                if (data.length && (options.autoSelect || options.autoselect)) {
                    data[0].selected = true;
                }

                // Save the unmatched value as a placeholder at the first output
                if (!this.active && value) {
                    placeholders[type] = value;
                }
            }

            // Add placeholder option
            if (options.placeholder) {
                data.unshift({
                    code: '',
                    address: placeholders[type],
                    selected: false
                });
            }

            $select.html(this.getList(data));
        },

        getList: function (data) {
            var list = [];

            $.each(data, function (i, n) {
                list.push(
                    '<option' +
                    ' value="' + (n.address && n.code ? n.address : '') + '"' +
                    ' data-code="' + (n.code || '') + '"' +
                    (n.selected ? ' selected' : '') +
                    '>' +
                    (n.address || '') +
                    '</option>'
                );
            });

            return list.join('');
        },

        reset: function (deep) {
            if (!deep) {
                this.output(menlei);
                this.output(dalei);
                this.output(zhonglei);
                this.output(xiaolei);
            } else if (this.$menlei) {
                this.$menlei.find(':first').prop('selected', true).trigger(EVENT_CHANGE);
            }
        },

        destroy: function () {
            this.unbind();
            this.$element.removeData(NAMESPACE);
        }
    };

    industrypicker.DEFAULTS = {
        autoSelect: true,
        placeholder: true,
        menlei: '—— 门类 ——',
        dalei: '—— 大类 ——',
        zhonglei: '—— 中类 ——',
        xiaolei: '—— 小类 ——'
    };

    industrypicker.setDefaults = function (options) {
        $.extend(industrypicker.DEFAULTS, options);
    };

    // Save the other industrypicker
    industrypicker.other = $.fn.industrypicker;

    // Register as jQuery plugin
    $.fn.industrypicker = function (option) {
        var args = [].slice.call(arguments, 1);

        return this.each(function () {
            var $this = $(this);
            var data = $this.data(NAMESPACE);
            var options;
            var fn;

            if (!data) {
                if (/destroy/.test(option)) {
                    return;
                }

                options = $.extend({}, $this.data(), $.isPlainObject(option) && option);
                $this.data(NAMESPACE, (data = new industrypicker(this, options)));
            }

            if (typeof option === 'string' && $.isFunction(fn = data[option])) {
                fn.apply(data, args);
            }
        });
    };

    $.fn.industrypicker.Constructor = industrypicker;
    $.fn.industrypicker.setDefaults = industrypicker.setDefaults;

    // No conflict
    $.fn.industrypicker.noConflict = function () {
        $.fn.industrypicker = industrypicker.other;
        return this;
    };

    $(function () {
        $('[data-toggle="industrypicker"]').industrypicker();
    });
});
